### PR TITLE
Architectural review

### PR DESCRIPTION
**Pull request by BennuFire**
Make neodash fast again.

### CODE REVIEW

**Comment by BennuFire** [comment]
- Stop using positional indexes and start using uuid ones instead
- Refactor data processing on sunburst
- Refactor data processing on treemap
- Refactor data processing on circle packing
- No re-execution of query after closing settings w/o changing it.

TODO : // Encapsulate data processing into an effect for Bars, Lines...

File: src/card/CardReducer.ts [hunk 1]
```diff
@@ -22,6 +23,7 @@ const update = (state, mutations) => Object.assign({}, state, mutations);
  */
 
 export const CARD_INITIAL_STATE = {
+  index: createUUID(),
```
Comment (nielsdejong) [inline src/card/CardReducer.ts:26]: i'd call it `id` instead of `index` here

Comment (nielsdejong) [reply src/card/CardReducer.ts:26]: Also refactor in `CardSelectors`, `CardThunks`

File: src/card/view/CardView.tsx [hunk 2]
```diff
@@ -130,7 +131,18 @@ const NeoCardView = ({
...
+
+  useEffect(() => {
+    if (!settingsOpen && queryChange) {
+      setLastRunTimestamp(Date.now());
+    }
+    setQueryChange(false);
+  }, [settingsOpen]);
+
+  useEffect(() => {
+    setQueryChange(true);
```
Comment (nielsdejong) [inline src/card/view/CardView.tsx:144]: Looks like this sometimes breaks queries that re-run on refresh rate

File: src/card/CardThunks.ts [hunk 3]
```diff
@@ -157,7 +157,7 @@ export const updateReportSettingThunk = (index, setting, value) => (dispatch: an
```
Comment (nielsdejong) [inline src/card/CardThunks.ts:157]: Generic comment:
Reloading a default dashboard with the new ID-based reports lets some reports stay in the state. (i.e. the state is never really cleaned up when doing a dashboard load).
Not sure where this happens or how to fix it though :)

File: src/page/Page.tsx [hunk 4]
```diff
@@ -26,7 +26,7 @@ export const NeoPage = ({
   isLoaded = true, // Whether the page is loaded and the cards can be displayed.
   onPageLayoutUpdate = () => {}, // action to take when the page layout is updated.
 }) => {
-  const getReportIndex = (pagenumber: number, index: number) => {
+  const getReportIndex = (pagenumber: number, index: number | string) => {
```
Comment (nielsdejong) [inline src/page/Page.tsx:29]: index should be `id` now, with `string` as the only allowed type

File: src/page/Page.tsx [hunk 5]
```diff
@@ -179,13 +180,13 @@ export const NeoPage = ({
...
       >
-        {reports.map((report, index) => {
+        {reports.map((report) => {
           const w = 12;
           // @ts-ignore
           return (
             <Grid
-              index={getReportIndex(pagenumber, index)}
-              key={getReportIndex(pagenumber, index)}
+              index={getReportIndex(pagenumber, report.index)}
```
Comment (nielsdejong) [inline src/page/Page.tsx:188]: maybe this is where it gets confused, as it's not sure how to handle the generated ID?

File: src/card/CardReducer.ts (hunk 1 above)
Comment (BennuFire) [reply src/card/CardReducer.ts]: Done

File: src/page/Page.tsx (hunk 4 above)
Comment (BennuFire) [reply src/page/Page.tsx]: done

File: src/page/Page.tsx (hunk 5 above)
Comment (BennuFire) [reply src/page/Page.tsx]: Nope, it was SettingsThunks L 20 (hardResetCardSettings dispatch)

File: src/card/view/CardView.tsx (hunk 2 above)
Comment (BennuFire) [reply src/card/view/CardView.tsx:144]: Can you elaborate on the sometimes? Im trying and its not failing :/

File: src/card/view/CardView.tsx [hunk 6]
```diff
@@ -130,7 +131,18 @@ const NeoCardView = ({
...
+
+  useEffect(() => {
+    if (!settingsOpen && selectorChange) {
+      setLastRunTimestamp(Date.now());
+    }
+    setSelectorChange(false);
+  }, [settingsOpen]);
+
+  useEffect(() => {
+    setSelectorChange(true);
```
Comment (nielsdejong) [inline src/card/view/CardView.tsx:144]: maybe we should explain here what `selectorChange` is.
I guess it means more something like "queryOrTypeChanged"?

File: src/dashboard/DashboardThunks.ts [hunk 7]
```diff
@@ -295,6 +309,15 @@ export const loadDatabaseListFromNeo4jThunk = (driver, callback) => (dispatch: a
 };
 
 export function upgradeDashboardVersion(dashboard: any, origin: string, target: string) {
+  if (origin == '2.2' && target == '2.3') {
+    dashboard.pages.forEach((p) => {
+      p.reports.forEach((r) => {
+        r.id = createUUID();
```
Comment (nielsdejong) [inline src/dashboard/DashboardThunks.ts:315]: remove index property?

File: src/dashboard/DashboardReducer.ts [hunk 8]
```diff
@@ -15,18 +15,20 @@ import {
...
   version: NEODASH_VERSION,
+
   settings: SETTINGS_INITIAL_STATE,
   pages: [FIRST_PAGE_INITIAL_STATE],
   parameters: {},
   extensions: {},
 };
 
 const update = (state, mutations) => Object.assign({}, state, mutations);
+console.log('TETST');
```
Comment (nielsdejong) [inline src/dashboard/DashboardReducer.ts:31]: Remove

File: src/page/Page.tsx [hunk 9]
```diff
@@ -26,7 +26,7 @@ export const NeoPage = ({
   isLoaded = true, // Whether the page is loaded and the cards can be displayed.
   onPageLayoutUpdate = () => {}, // action to take when the page layout is updated.
 }) => {
-  const getReportIndex = (pagenumber: number, index: number) => {
+  const getReportIndex = (pagenumber: number, index: string) => {
```
Comment (nielsdejong) [inline src/page/Page.tsx:29]: should now be `getReportId` i think? and change parameter `index` to `id`?

File: src/page/Page.tsx [hunk 10]
```diff
@@ -46,7 +46,7 @@ export const NeoPage = ({
   const loadingMessage = <div>Loading card...</div>;
   const [isDragging, setIsDragging] = React.useState(false);
   const [layouts, setLayouts] = React.useState(defaultLayouts);
-  const [lastElement, setLastElement] = React.useState(<div key={getReportIndex(pagenumber, 999999)}></div>);
+  const [lastElement, setLastElement] = React.useState(<div key={getReportIndex(pagenumber, '999999')}></div>);
```
Comment (nielsdejong) [inline src/page/Page.tsx:49]: i don't think the concept of lastElement makes sense anymore

File: src/page/PageReducer.ts [hunk 11]
```diff
@@ -57,13 +60,15 @@ export const pageReducer = (state = PAGE_INITIAL_STATE, action: { type: any; pay
...
   if (action.type.startsWith('PAGE/CARD/')) {
-    const { index } = payload;
+    const { id } = payload;
     return {
       ...state,
       reports: [
-        ...state.reports.slice(0, index),
-        cardReducer(state.reports[index], action),
-        ...state.reports.slice(index + 1),
+        ...state.reports.filter((o) => o.id !== id),
```
Comment (nielsdejong) [inline src/page/PageReducer.ts:67]: will this shuffle the array on every edit? moving the edited to the end?

File: src/page/PageThunks.ts [hunk 12]
```diff
@@ -19,21 +20,21 @@ export const addReportThunk =
     }
   };
 
-export const removeReportThunk = (index: number) => (dispatch: any, getState: any) => {
+export const removeReportThunk = (id: number) => (dispatch: any, getState: any) => {
```
Comment (nielsdejong) [inline src/page/PageThunks.ts:23]: id: string

File: src/card/CardSelectors.ts [hunk 13]
```diff
@@ -1,6 +1,6 @@
 export const getDashboardTitle = (state: any) => state.dashboard.title;
 
-export const getReportState = (state: any, index: any) => {
+export const getReportState = (state: any, id: any) => {
   const { pagenumber } = state.dashboard.settings;
-  return state.dashboard.pages[pagenumber].reports[index];
+  return state.dashboard.pages[pagenumber].reports.find((o) => o.id === id);
```
Comment (nielsdejong) [inline src/card/CardSelectors.ts:5]: i have a crazy idea here but you will hate me

File: src/card/view/CardView.tsx (hunk 6 above)
Comment (nielsdejong) [reply src/card/view/CardView.tsx:144]: there's now an issue where queries don't refresh if you open/close the settings.
E.g. `return rand()`, do we want that intentionally?

Comment (BennuFire) [reply src/card/view/CardView.tsx:144]: 1. Comment added. 
2. We should encourage the use of the refresh button. We can set it on by default.

File: src/dashboard/DashboardThunks.ts (hunk 7 above)
Comment (BennuFire) [reply src/dashboard/DashboardThunks.ts:315]: They never had it.

File: src/dashboard/DashboardReducer.ts (hunk 8 above)
Comment (BennuFire) [reply src/dashboard/DashboardReducer.ts:31]: What?

File: src/page/Page.tsx (hunk 9 above)
Comment (BennuFire) [reply src/page/Page.tsx:29]: Used only for the grid keys

File: src/page/Page.tsx (hunk 10 above)
Comment (BennuFire) [reply src/page/Page.tsx:49]: Same as usual

File: src/card/CardSelectors.ts (hunk 13 above)
Comment (BennuFire) [reply src/card/CardSelectors.ts:5]: Not crazy.... soon to be implemented.

File: src/page/PageThunks.ts (hunk 12 above)
Comment (BennuFire) [reply src/page/PageThunks.ts:23]: Solved

**Review by nielsdejong — APPROVED** [review]

**Pull request merged**